vamb/reclustering.py

```diff
@@ -0,0 +1,319 @@
+            r = list(sel.query('gene == @c')['contig'])
+            r.sort()
+            return r
+
+
+        if multi_mode:
+            data['bin'] = data['orf'].str.split(pat='.', n=0, expand=True)[0]
+            counts = data.groupby(['bin', 'gene'])['orf'].count()
+            res = {}
+            for b,vs in counts.groupby(level=0):
+                cs = extract_seeds(vs.droplevel(0), data.query('bin == @b', local_dict={'b':b}))
+                res[b] = [c.split('.',1)[1] for c in cs]
+            return res
+        else:
+            counts = data.groupby('gene')['orf'].count()
+            return extract_seeds(counts, data)
+
+
+def run_prodigal(fasta_path, num_process, output):
+    contigs = {}
+    for h, seq in fasta_iter(fasta_path):
+        contigs[h] = seq
+
+    total_len = sum(len(s) for s in contigs.values())
+    split_len = total_len // num_process
+
+    cur = split_len + 1
+    next_ix = 0
+    out = None
+    with contextlib.ExitStack() as stack:
+        for h,seq in contigs.items():
+            if cur > split_len and next_ix < num_process:
+                if out is not None:
+                    out.close()
+                out = open(os.path.join(output, 'contig_{}.fa'.format(next_ix)), 'wt')
+                out = stack.enter_context(out)
+
+                cur = 0
+                next_ix += 1
+            out.write(f'>{h}\n{seq}\n')
+            cur += len(seq)
+
+    try:
+        process = []
+        for index in range(next_ix):
+            with open(os.path.join(output, f'contig_{index}_log.txt') + '.out', 'w') as prodigal_out_log:
+                p = subprocess.Popen(
+                        ['prodigal',
+                         '-i', os.path.join(output, f'contig_{index}.fa'),
+                         '-p', 'meta',
+                         '-q',
+                         '-m', # See https://github.com/BigDataBiology/SemiBin/issues/87
+                         '-a', os.path.join(output, f'contig_{index}.faa')
+                         ],
+                    stdout=prodigal_out_log)
+                process.append(p)
+
+        for p in process:
+            p.wait()
+
+    except:
+        sys.stderr.write(
+            f"Error: Running prodigal fail\n")
+        sys.exit(1)
+
+    contig_output = os.path.join(output, 'contigs.faa')
+    with open(contig_output, 'w') as f:
+        for index in range(next_ix):
+            f.write(open(os.path.join(output, 'contig_{}.faa'.format(index)), 'r').read())
+    return contig_output
+
+
+def cal_num_bins(fasta_path, binned_length, num_process, multi_mode=False, output = None, orf_finder = 'prodigal'):
+    '''Estimate number of bins from a FASTA file
+
+    Parameters
+    fasta_path: path
+    binned_length: int (minimal contig length)
+    num_process: int (number of CPUs to use)
+    multi_mode: bool, optional (if True, treat input as resulting from concatenating multiple files)
+    '''
+    with tempfile.TemporaryDirectory() as tdir:
```

Comment:
No big deal, but can we make it an argument where it writes its temp files? We might want to put it all in a single tmp dir where the user specified.

Reply:
agree, those files take long to generate and could be re-used. The reason the temp folder sometimes is more convenient, is that the contig names are concatenated with bin labels somewhere along the lines and make a mess if a different cluster file is used. Can be refactored though